tooling/debugger/src/context.rs

```diff
@@ -21,6 +21,128 @@ use thiserror::Error;
+pub struct AddressMap {
+    // A Noir program is composed by
+    // `n` ACIR circuits
+    //       |_ `m` ACIR opcodes
+    //                |_ Acir call
+    //                |_ Acir Brillig function invocation
+    //                           |_ `p` Brillig opcodes
+    //
+    // We need to inline such structure to a one-dimension "virtual address" space
+    // We define the address space as a contiguous space where all ACIR and
+    // Brillig opcodes from all circuits are laid out
+    //
+    // `addresses: Vec<Vec<usize>>``
+    //  * The first vec is `n` sized - one element per ACIR circuit
+    //  * Each nested vec is `m` sized - one element per ACIR opcode in circuit
+    //    * Each element is the "virtual address" of such opcode
+    //
+    // For flattening we map each ACIR circuit and ACIR opcode with a sequential address number
+    // We start by assigning 0 to the very first ACIR opcode and then start accumulating by
+    // traversing by depth-first
+    //
+    //  * an ACIR call accumulates 1 (since it's a single opcode)
+    //  * an ACIR Brillig call accumulates as many opcodes the brillig function has (`p`)
+    //
+    // As a result the flattened addresses list may have "holes".
+    // This holes are a side effect of the brillig function calls
+    //
+    addresses: Vec<Vec<usize>>,
+
+    // virtual address of the last opcode of the program
+    last_valid_address: usize,
+}
+
+impl AddressMap {
+    pub(super) fn new(
+        circuits: &[Circuit<FieldElement>],
+        unconstrained_functions: &[BrilligBytecode<FieldElement>],
+    ) -> Self {
+        let mut result = Vec::with_capacity(circuits.len());
+        let mut circuit_address_start = 0usize;
+
+        for circuit in circuits {
+            let mut circuit_addresses = Vec::with_capacity(circuit.opcodes.len());
+            // push the starting address of the first opcode
+            circuit_addresses.push(circuit_address_start);
+
+            // iterate opcodes _init_ (all but last)
+            let last_opcode_address = circuit.opcodes.iter().take(circuit.opcodes.len() - 1).fold(
+                circuit_address_start,
+                |acc, opcode| {
+                    let acc = acc + Self::calculate_opcode_size(opcode, unconstrained_functions);
+                    // push the starting address of the next opcode
+                    circuit_addresses.push(acc);
+                    acc
+                },
+            );
+            // last opcode size should be next circuit address start
+            let last_opcode = circuit.opcodes.last().expect("There is a last opcode");
+            circuit_address_start = last_opcode_address
+                + Self::calculate_opcode_size(last_opcode, unconstrained_functions);
+
+            result.push(circuit_addresses);
+        }
+        let last_valid_address = circuit_address_start - 1;
+        Self { addresses: result, last_valid_address }
```

Comment:
I think this code is simpler to reason about in an imperative style (also using a closure for the opcode size computation):

```rust
let opcode_address_size = |opcode| {
  if let Opcode::BrilligCall { id, .. } = opcode {
    unconstrained_functions[*id as usize].bytecode.len()
  } else {
    1
  }
};

let addresses = Vec::with_capacity(circuits.len());
let mut address = 0usize;

for circuit in circuits {
  let mut circuit_addresses = Vec::with_capacity(circuit.opcodes.len());
  for opcode in &circuit.opcodes {
    circuit_addresses.push(address);
    address += opcode_address_size(opcode);
  }
  addresses.push(circuit_addresses);
}

Self { addresses, last_valid_address: address - 1 }
```